src/Workspaces/CSharp/Portable/Extensions/CastExpressionSyntaxExtensions.cs

```diff
@@ -540,5 +555,22 @@ private static bool IsRequiredImplicitNumericConversion(ITypeSymbol sourceType,
                     return false;
             }
         }
+
+        private static bool IsInUncheckedContext(CastExpressionSyntax cast)
+        {
+            SyntaxNode currentNode = cast;
+
+            do
+            {
+                if (currentNode.Kind() == SyntaxKind.UncheckedExpression || currentNode.Kind() == SyntaxKind.UncheckedStatement)
+                {
+                    return true;
+                }
+
+                currentNode = currentNode.Parent;
+            } while (currentNode is ExpressionSyntax || currentNode is StatementSyntax);
```

Comment:
Lambda in unchecked expression: Is that an important scenario? I'd prefer to still err on the conservative side.

Checked expression/statement: Indeed we should bail out and return false, I'll fix that and also add a test.